addons/print_docsaway/models/print_docsaway.py

```diff
@@ -0,0 +1,391 @@
+
+    # station fields
+    docsaway_station = fields.Char('Docsaway Station', help="Docsaway Station ID.")
+    docsaway_courier = fields.Char('Docsaway Courier', help="Docsaway Printer ID.")
+    docsaway_zone = fields.Integer('Docsaway Zone', help="Docsaway Station ID.")
+    # transaction fields
+    docsaway_reference = fields.Char('Docsaway Reference', help="Docsaway Reference Order.")
+    docsaway_approved = fields.Char('Docsaway Approved', help="Docsaway Transation Status.")
+
+
+    def _docsaway_prepare_printing(self):
+        """ Pre-process the sending by computing the best station for each given orders.
+        """
+        # minimize the number of call to DocsAway API by grouping by partner
+        # (for localization) and by ink to find the best station.
+        stations = {}
+        for order in self:
+            try:
+                if not (order.partner_id.id, order.ink) in stations:
+                    stations[(order.partner_id.id, order.ink)] = order.provider_id._docsaway_fetch_station(order.partner_id.country_id.code, order.ink, order.paper_weight)
+                response = stations[(order.partner_id.id, order.ink)]
+                order.write({
+                    'docsaway_station' : response['stationAuto']['stationID'],
+                    'docsaway_courier' : response['stationAuto']['courierID'],
+                    'docsaway_zone' : response['stationAuto']['zone'],
+                    'state' : 'ready_to_send'
+                })
+            except (DocsawayException, Warning) as e:
+                order.write({
+                    'state' : 'error',
+                    'error_message' : e,
+                })
+
+    def _docsaway_deliver_printing(self):
+        """ Send the orders for delivery to DocsAway.
+            self is a recordset of print.order having the same provider type
+        """
+        provider_balance = {}
+        for order in self:
+            if order.state == 'ready_to_send':
+                # create params to send order to docsaway.
+                address = {
+                    'name' : order.partner_name,
+                    'street' : order.partner_street,
+                    'street2' : order.partner_street2,
+                    'state' : order.partner_state_id.name,
+                    'city' : order.partner_city,
+                    'zip' : order.partner_zip,
+                    'country_code' : order.partner_country_id.code
+                }
+                printing_infos = {
+                    'station_id' : order.docsaway_station,
+                    'courier_id' : order.docsaway_courier,
+                    'zone' : order.docsaway_zone,
+                    'ink' : order.ink,
+                    'paper' : order.paper_weight
+                }
+                try:
+                    response = order.provider_id._docsaway_send_document(address, printing_infos, order.attachment_id.datas)
+                    order.write({
```

Comment:
I would move the write outside of the try / except statement, as it should not be in the same protection (issue in the write for example). The try/catch could forge a result, and the result would be then used to update the order (same remark as above, note).